linkerd/app/integration/src/tests/direct.rs

```diff
@@ -0,0 +1,134 @@
+
+    let (inbound, _profile_in) = {
+        let (proxy, profile) = mk_inbound(srv, in_identity.service(), &dst).await;
+        let proxy = proxy.run_with_test_env(in_identity.env).await;
+        (proxy, profile)
+    };
+
+    let (outbound, _profile_out, _dst) = {
+        let ctrl = controller::new();
+        let dst = ctrl.destination_tx(dst);
+        dst.send(
+            controller::DestinationBuilder {
+                hint: controller::Hint::H2,
+                opaque_port: Some(inbound.inbound.port()),
+                identity: Some(in_svc_acct.to_string()),
+                ..Default::default()
+            }
+            .into_update(srv_addr),
+        );
+        let (proxy, profile) = mk_outbound(srv_addr, ctrl, out_identity).await;
+        (proxy, profile, dst)
+    };
+
+    let client = client::http1(outbound.outbound, "opaque.test.svc.cluster.local");
+
+    assert_eq!(client.get("/").await, "hello");
+}
+
+/// Reproduces linkerd/linkerd2#9888. A proxy receives HTTP traffic direct
+/// traffic with a transport header, and the port is also in the
+/// `INBOUND_PORTS_DISABLE_PROTOCOL_DETECTION` env var.
+#[tokio::test]
+async fn opaque_hinted() {
+    let _trace = trace_init();
+
+    // identity is always required for direct connections
+    let in_svc_acct = "foo.ns1.serviceaccount.identity.linkerd.cluster.local";
+    let in_identity = identity::Identity::new("foo-ns1", in_svc_acct.to_string());
+
+    let out_svc_acct = "bar.ns1.serviceaccount.identity.linkerd.cluster.local";
+    let out_identity = identity::Identity::new("bar-ns1", out_svc_acct.to_string());
+
+    let srv = server::http1().route("/", "hello").run().await;
+    let srv_addr = srv.addr;
+    let dst = format!("opaque.test.svc.cluster.local:{}", srv_addr.port());
+
+    let (inbound, _profile_in) = {
+        let id_svc = in_identity.service();
+        let mut env = in_identity.env;
+        env.put(
+            app::env::ENV_INBOUND_PORTS_DISABLE_PROTOCOL_DETECTION,
+            srv_addr.port().to_string(),
+        );
+        let (proxy, profile) = mk_inbound(srv, id_svc, &dst).await;
+        let proxy = proxy.run_with_test_env(env).await;
+        (proxy, profile)
+    };
+
+    let (outbound, _profile_out, _dst) = {
+        let ctrl = controller::new();
+        let dst = ctrl.destination_tx(dst);
+        dst.send(
+            controller::DestinationBuilder {
+                hint: controller::Hint::Opaque,
+                opaque_port: Some(inbound.inbound.port()),
+                identity: Some(in_svc_acct.to_string()),
+                ..Default::default()
+            }
+            .into_update(srv_addr),
+        );
+        let (proxy, profile) = mk_outbound(srv_addr, ctrl, out_identity).await;
+        (proxy, profile, dst)
+    };
+
+    let client = client::http1(outbound.outbound, "opaque.test.svc.cluster.local");
+
+    assert_eq!(client.get("/").await, "hello");
```

Comment:
do you think it's worth adding an assert here to check the status code? I'm thinking it might give some visibility if the should ever fail in the future? Then again, I think what's in here is already good 🤷🏻

Reply:
The `client.get(...)` method panics if the status isn't 200 OK. We would use `client.request(...)` to not fail on other status codes. So, this already tests that it's OK. :)